### PR TITLE
ENH: Compressed layout for fixed-aspect axes

### PR DESCRIPTION
**Pull request by jklymak**
## PR Summary

EDIT 5 May: new write up here...

Adds a new layout option `compress_layout=True` to `Figure.figure`. This option attempts to compress axes if they are too far apart because of fixed aspect ratios....

This is most feature complete with `constrained_layout=True`, but works if folks don't use colorbars for no layout or `tight_layout`.  The real issue is that we cannot handle nested gridspecs with this, and colorbars, by default, are nested gridspecs.  (Don't get me started on *how* we make colorbars - I think that needs a rethink).  

See new example: https://36260-1385122-gh.circle-artifacts.com/0/doc/build/html/gallery/subplots_axes_and_figures/compress_axes.html#sphx-glr-gallery-subplots-axes-and-figures-compress-axes-py

### Before:

```python
fig, axs = plt.subplots(2, 2, figsize=(5, 3), facecolor='0.75',
                        sharex=True, sharey=True, constrained_layout=True)
for ax in axs.flat:
    ax.set_aspect(1.0)
plt.show()
```
OldCL

### After
```python
fig, axs = plt.subplots(2, 2, figsize=(5, 3), facecolor='0.75',
                        sharex=True, sharey=True, constrained_layout=True, 
                        compress_layout=True)
for ax in axs.flat:
    ax.set_aspect(1.0)
plt.show()
```

NewCL

## PR Checklist

- [x] suptitle!

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

**Comment by jklymak** [comment]
Remilestoned for 3.4, but would be nice to get some comment for this.  Is a string value for the `constrained_layout`  kwarg OK?  "squish" OK or does someone have a better name?  

I think it would be *possible* for the just to be run all the time when CL is turned on; its quite fast, but I've not tried every weird combo of axes yet. 

I also think it would be possible to run this stand-alone.  `_squish` doesn't actually rely on anything in CL.  Its nice though to have it injected at the draw stage because otherwise the user has to trigger a draw manually.

**Comment by anntzer** [comment]
Are the axes not squished horizontally because of the long xlabels?  If so it may be nice to provide an example where squishing occurs in both directions.

**Comment by jklymak** [comment]
@anntzer Not quite sure what you mean.  The aspect ratio of the figure is such that squishing only happens in x in this case, but if the aspect ratios were swapped the squishing happens in x.  

If you mean the spacing is different between the axes in x and y, then yes, thats because of the large x-labels.  The point of constrained_layout is non-overlapping labels, so thats a feature, not a bug.

If instead we want a layout where labels are allowed to overlap and the plot positions are simply a fixed distance apart, then that is really a lot easier than what is done here since you can do all the positioning with the axes positions and don't need to know about the size of the decorations.

File: lib/matplotlib/figure.py [hunk 1]
```diff
@@ -2406,6 +2410,47 @@ def init_layoutbox(self):
...
+        bboxes = None
+        if self.axes:
+            adjusted = False
+            if self.get_tight_layout():
+                try:
+                    w_pad, h_pad = self.tight_layout(
+                            **self._tight_parameters)
+                    # convert pads to inches...
+                    fs = FontProperties(size=mpl.rcParams["font.size"])
+                    fs = fs.get_size_in_points() / 72
```
Comment (jklymak) [inline lib/matplotlib/figure.py:2427]: This flake8 error looks spurious?

**Comment by jklymak** [comment]
I'm closing this, not because I think its a bad idea, but because it needs to be substantially revised with the new CL changes.

**Pull request closed without merging**